### PR TITLE
21332 Dissolutions Job - Implement stage 3

### PR DESCRIPTION
**Pull request by leodube-aot**
*Issue #:* /bcgov/entity#21332

*Description of changes:*
- Added logic to dissolutions job to move businesses to stage 3. For detailed logic, reference [tech design](https://community.inkdrop.app/note/a27b7a79c8cdf7db0ab19be10b4fc2e8/note:Ej0jewYQK#user-content-3-processing-actual-dissolution-of-businesses)
- Added unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py [hunk 1]
```diff
@@ -73,7 +77,46 @@ def shell_context():
...
+    batches = db.session.query(Batch).filter(Batch.id.in_(subquery)).all()
+
+    for batch in batches:
+        batch.status = Batch.BatchStatus.COMPLETED
+        batch.end_date = datetime.utcnow()
+        batch.save()
+
+
+def stage_1_process(app: Flask):  # pylint: disable=redefined-outer-name,too-many-locals 
     """Initiate dissolution process for new businesses that meet dissolution criteria."""
```
Comment (leodube-aot) [inline jobs/involuntary-dissolutions/involuntary_dissolutions.py:120]: Renaming to be consistent with the other stage processing functions. Also updated some of the description strings to be more specific as to what each stage does.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py [hunk 2]
```diff
@@ -73,7 +77,46 @@ def shell_context():
...
 
 
-def initiate_dissolution_process(app: Flask):  # pylint: disable=redefined-outer-name,too-many-locals
+def create_invountary_dissolution_filing(business_id: int):
+    """Create a filing entry to represent an involuntary dissolution filing."""
+    filing = Filing()
+    filing.business_id = business_id
+    filing.filing_type = CoreFiling.FilingTypes.DISSOLUTION
+    filing.filing_sub_type = DissolutionTypes.INVOLUNTARY
+    filing.save()
```
Comment (leodube-aot) [inline jobs/involuntary-dissolutions/involuntary_dissolutions.py:86]: This is the simplest filing for involuntary dissolution I could think of. Not sure if we need to include more information in the filing or if this is okay for now.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py [hunk 3]
```diff
@@ -73,7 +78,55 @@ def shell_context():
...
+        # mark any failure for human review
+        capture_message(
+            f'Queue Error: Failed to place filing {filing_id} on Queue with error:{err}',
+            level='error'
+        )
+
+
+def mark_eligible_batches_completed():
+    """Mark batches completed if all of their associated batch_processings are compeleted."""
+    AliasBatchProcessing = aliased(BatchProcessing)  # pylint: disable=invalid-name # noqa N806
```
Comment (leodube-aot) [inline jobs/involuntary-dissolutions/involuntary_dissolutions.py:108]: Had to alias `BatchProcessing` to get this query to work. The sql query is a bit weird but it was the only way I could figure out how to select batches where their associated batch_processings entries have status COMPLETED or WITHDRAWN.

Screenshot 2024-06-20 at 12 11 39 PM

**Comment by leodube-aot** [comment]
After running stage 3, this is what the database entries look like:

`batches`:
Screenshot 2024-06-21 at 8 04 30 AM

`batch_processing`:
Screenshot 2024-06-21 at 8 05 13 AM

`filings`:
Screenshot 2024-06-21 at 8 08 39 AM

Still working on getting it working with the filer, will update with the `businesses` entries.

**Comment by chenhongjing** [comment]
> After running stage 3, this is what the database entries look like:
> 
> `batches`: Screenshot 2024-06-21 at 8 04 30 AM
> 
> `batch_processing`: Screenshot 2024-06-21 at 8 05 13 AM
> 
> `filings`: Screenshot 2024-06-21 at 8 08 39 AM
> 
> Still working on getting these changes working with the filer, I think it's just a local setup issue. We know from a previous ticket that involuntary dissolution filings can be processed by the filer. There may be some small changes needed in creating the involuntary dissolution filing but while I work out the connection issues this ticket is good to start getting reviews.
> 
> I'll also update with the `businesses` table entries.

emmmm...not sure if your local filer works well and you need to re-setup filer with Python 3.9+. When I resolved conflicts after recent filer updates, I had to do it.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 2 above)
Comment (argush3) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:89]: We'll need to populated `filing.filing_json` for sure.  The dissolution processor looks at it.

Will take a look and see if we need anything else populated in the `filings` table.

**Comment by leodube-aot** [comment]
> emmmm...not sure if your local filer works well and you need to re-setup filer with Python 3.9+. When I resolved conflicts after recent filer updates, I had to do it.

It ended up being just a problem with loading my `.env` file. For some reason `os.getenv('<VARIABLE>', '<fallback>')` is only returning the fallback value, even if the VARIABLE is set correctly in the `.env` file (for example, `os.getenv('NATS_FILER_SUBJECT', 'error')` was only returning 'error' even though the correct value of `entity.filer` is set in the environment variable file). I had to manually set the VARIABLE value as the fallback value to get it to work.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 2 above)
Comment (leodube-aot) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:110]: Yeah you're right, the `filing_json` was needed for the entity filer to process it, and it seems to be working now. Other than that, let me know if there is anything else we should be adding.

Comment (argush3) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:110]: was looking at an admin dissolution filing and I think populating the `filing_date` and `source` field would be good too.

Example admin dissolution filing in dev:
image

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 3 above)
Comment (argush3) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:131]: does something like this work?

```
batches = (
        db.session.query(Batch)
        .join(BatchProcessing, Batch.id == BatchProcessing.batch_id)
        .filter(Batch.batch_type == 'INVOLUNTARY_DISSOLUTION')
        .filter(Batch.status != 'COMPLETED')
        .group_by(Batch.id)
        .having(
            ~db.session.query(BatchProcessing.id)
            .filter(BatchProcessing.batch_id == Batch.id)
            .filter(BatchProcessing.status.notin_(['COMPLETED', 'WITHDRAWN']))
            .exists()
        )
        .all()
    )
```

File: jobs/involuntary-dissolutions/tests/unit/test_involuntary_dissolutions.py [hunk 4]
```diff
@@ -217,3 +278,44 @@ def test_stage_2_process_update_business(app, session, test_name, status, step):
...
+        trigger_date=TRIGGER_DATE
+    )
+
+    if test_name == 'MOVE_BACK_TO_GOOD_STANDING':
+        business.last_ar_date = datetime.utcnow()
+        business.save()
+
+    with patch('involuntary_dissolutions.put_filing_on_queue'):
+        qsm = MagicMock()
+        await stage_3_process(app, qsm)
```
Comment (argush3) [inline jobs/involuntary-dissolutions/tests/unit/test_involuntary_dissolutions.py:316]: can we have an assertion check that verifies that `put_on_filing_queue` is called for the `DISSOLVE_BUSINESS` test?

**Review by argush3 — APPROVED** [review]
Looks good.

Pre-approving with assumption my 2 outstanding comments will be addressed.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py [hunk 5]
```diff
@@ -73,7 +78,78 @@ def shell_context():
...
-def initiate_dissolution_process(app: Flask):  # pylint: disable=redefined-outer-name,too-many-locals
+def create_invountary_dissolution_filing(business_id: int):
+    """Create a filing entry to represent an involuntary dissolution filing."""
+    business = Business.find_by_internal_id(business_id)
+
+    filing = Filing()
+    filing.business_id = business.id
+    filing._filing_type = CoreFiling.FilingTypes.DISSOLUTION  # pylint: disable=protected-access
+    filing._filing_sub_type = DissolutionTypes.INVOLUNTARY  # pylint: disable=protected-access
+    filing._status = Filing.Status.PAID  # pylint: disable=protected-access
```
Comment (chenhongjing) [inline jobs/involuntary-dissolutions/involuntary_dissolutions.py:89]: Maybe no need to set status for filing? 💭 🤔  Just some observation, the [listener](https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/models/filing.py#L1033) will check and update the status before inserting. This is also shown in your testing results that the status of filing is DRAFT before filer processes it.

**Review by chenhongjing — COMMENTED** [review]
LGTM 🐱  Will approve it after the outstanding comments (mainly the one from Argus) are addressed.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 3 above)
Comment (leodube-aot) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:131]: Unfortunately no, it doesn't. I get `Select statement '<sqlalchemy.sql.selectable.Select object at 0x107e74fd0>' returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually`, which is the issue I encountered initially.

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 5 above)
Comment (leodube-aot) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py]: Good point, updated!

File: jobs/involuntary-dissolutions/involuntary_dissolutions.py (hunk 3 above)
Comment (leodube-aot) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:130]: Unfortunately no, it doesn't. I get `Select statement '<sqlalchemy.sql.selectable.Select object at 0x107e74fd0>' returned no FROM clauses due to auto-correlation; specify correlate(<tables>) to control correlation manually`, which is the issue I encountered initially.

Comment (argush3) [reply jobs/involuntary-dissolutions/involuntary_dissolutions.py:130]: ok, resolving.

**Review by chenhongjing — APPROVED** [review]

**Pull request merged**